### ISSUE TITLE
Cannot open passage editor using keyboard alone

### ISSUE THREAD

**Issue by klembot** [klembot/twinejs#565]
See #564

**Comment by HiEv** [comment]
The TAB key is very difficult to use in the passage editor.  If the cursor is on the passage name line, then you need to hit TAB _five_ times before you get to editing the passage text, and I have no idea where the focus is on the intervening four TABs.  Also, once you do tab to the passage text, the scrollbar doesn't move to show where the cursor currently is.

There is no easy way to use the TAB to select the items on the bottom menu bars, and ALT should toggle activating/deactivating the top menu bar.  The TAB key should highlight the item you're currently on, then you could use the ENTER or SPACE keys to activate items, plus the UP/DOWN arrow keys to display/navigate the item's menu (if it has one).  Currently the "Home" icon is the only one that indicates when it has focus.

On the main window, the arrow keys should let you select (highlight) the story and then both ENTER or SPACE should let you open that story.  Similarly, in the story window it would be nice if it let you move around the passages that way.  Possibly also some key could pop out a "Search" you could use to find all stories/passages with names that contain whatever you typed.

There are also no keyboard shortcuts for several things.  We need keyboard shortcuts for:
- Closing the passage or story windows (should be CTRL+W)
- Opening the JavaScript window
- Opening the Stylesheet window
- Toggling "Snap to grid" (with an indicator of state)
- Quick find (should be CTRL+F)
- Play
- Test
- Add passage
- possibly any other things accessible through the bottom menu bar

Also, as mentioned in the item this is linked from, the "alt" attribute should be set for all buttons and the like for accessibility.

Finally, lots of useful information on making websites accessible can be found at:
- [Web Content Accessibility Guidelines](https://www.w3.org/WAI/standards-guidelines/wcag/) (WCAG)
- [Section508](https://www.section508.gov/test/web-software) Government-wide IT Accessibility Program
- [WebAIM](https://webaim.org/) (Accessibility In Mind) site.
- [SortSite - Accessibility Checker and Validator](https://www.powermapper.com/products/sortsite/checks/accessibility-checks/)

**Comment by klembot** [comment]
@HiEv, let's please keep this issue scoped to the accessibility of the passage editor... many of the things you've mentioned are outside that scope. They sound valid but I really want to avoid a single issue that tries to cover everything under the sun, so that the work can be divided up and steady progress can be made across releases. Opening more issues would be welcomed.

**Comment by HiEv** [comment]
Except for the addendum at the end on general accessibility recommendations, most of that seems to fit the category of being able to access a passage to edit it using keyboard navigation alone to me.

If, for example, you opened Twine and wanted to edit the JavaScript passage of a story using the keyboard alone, then most of what I described above would come into play.

If you'd like to break that up into separate issues, feel free, but I'm unclear where those breaks should occur or how spreading these related issues out improves things, so I wouldn't know where to begin doing that myself.

**Comment by greyelf** [comment]
Any keyboard (gesture for touch?) short-cuts chosen would at least need to cover all the different Operating Systems (device types?) in common use, which means that CTRL or ALT can't be the only modification key used to access extended key functionality.

**Comment by klembot** [comment]
I retitled this issue to make its scope clearer; point is, my understanding right now is that users of screen readers are unable to open the passage editor at all. Getting this fixed needs to be prioritized over other things mentioned in the comment above, which would make the keyboard-only experience better.

**Comment by HiEv** [comment]
Since there may be dozens or hundreds of nodes in a story, some sort of keyboard-navigable "story node search" needs to be implemented.  The best version would probably be a textbox (labeled "Story Node Search") with an attached dropdown list that fills with all node names with matching text in the name, sorted alphabetically.  This could be placed on the bottom menu-bar between the "Story Name" menu and the "Quick Find" textbox.  Matching node names could also be highlighted above as well, like what happens when using the "Quick Find".

The UpArrow and DownArrow keys should show different items in the dropdown list without clearing the search list, and hitting the Enter key should open up the corresponding node, or the first node if the arrow keys weren't used to select from the dropdown.  If no matching node names were found, a notification that "no matching node name was found" upon hitting Enter should be given.

**Comment by tangela19** [comment]
Greetings, 

After trying the twine editor recently, both on the web as well as the windows app, I'm sorry to say that this still seems to be in the same state as when this post was made. I'm respectfully curious whether there are any internal efforts being made regarding blind accessibility in general and keyboard navigation in specific.

**Comment by klembot** [comment]
I would like to address this in Twine 2.4. It is not ready to be tested, but I would really appreciate feedback at the alpha stage on accessibility issues.

**Comment by tangela19** [comment]
Hello, I would love to test and provide feedback when this becomes ready
for public consumption. Thanks for working on it.

**Comment by klembot** [comment]
2.4 improves keyboard accessibility, but it's very likely that there's ground still to be covered. I'm closing this issue, but please feel free to open issues based on the changes in 2.4.

**Comment by tangela19** [comment]
Is there any documentation of what has been changed, and what we should be able to expect to work now?

**Comment by klembot** [comment]
Broadly speaking, you should now be able to navigate the story map and select passages using the keyboard or assistive technology. The same should be true of the story list. There aren't any new keyboard shortcuts added, but it's something people have asked for in other issues.
